Pstl

alexstrel opened the pull request:
Added C++17 PSTL version. Tested on x86 (gcc-10.2 + TBB 2020.3) and nvidia march (nvc++ 2021.3)

alexstrel committed:
original commit

alexstrel committed:
finished pstl version, not tested yet

alexstrel committed:
fixed compiler errors

alexstrel committed:
uncommented conversion routine

alexstrel committed:
minor cleaning

alexstrel committed:
some minor changes in the compute kernels

alexstrel committed:
fixed post-processing stage

alexstrel committed:
fixed bug in KalmanUpdate

alexstrel committed:
Merge branch 'main' into pstl

alexstrel committed:
added warm-up stage

kakwok commented:
@alexstrel Can you add the commands you used to compile the code? I can add the support to the build script if needed

alexstrel committed:
some modification of build.py, not finished yet

alexstrel commented:
@kakwok made initial commit, not tested. pstl technology will refer to many targets, need  to adjust the script, maybe add "backend" option? BTW, made std=c++17 as a default option (as it is for the recently released  gcc-11.1.0, dpcpp and nvc++ )

alexstrel committed:
added custom counting iterator, removed explicit references
to tbb and thrust namspaces (no longer needed). Added compilation
instruction for multicore for nvc++

alexstrel committed:
simplified allocators

alexstrel committed:
added compilation options, slightly improved performance

kakwok commented on build.py:
```diff
@@ -27,7 +27,11 @@
+    },
+    "pstl":{
+        "pstl":['gcc','nvc++']
+    },
```
@alexstrel We meant to support different backend like this: (see a few lines below for the comment on kokkos)
``` 
   "pstl":{
        "cpu":['gcc'] # add other compilers
        'cuda': ['nvc++']
    },
```

cerati commented:
@kakwok @alexstrel is this PR ready to be merged?

alexstrel commented:
@cerati  the branch was now checked against nvc++ (21.7)  ( both gpu and x86) as well as g++ (v. 11.2.0)  + tbb for x86. So far no issues. I need to check build script and README for the merge.

kakwok committed:
Add instructions for cori

alexstrel committed:
Merge branch 'main' into pstl

kakwok committed:
update for V100 compute capability

alexstrel committed:
Merge remote-tracking branch 'upstream/pstl' into pstl

alexstrel committed:
fixed a typo in the function name

alexstrel committed:
added oneapi dpl version (for Intel Xe targets)

kakwok committed:
Merge branch 'main' into pstl

kakwok commented:
@alexstrel @cerati I fixed the conflict in the build script in https://github.com/cerati/p2r-tests/pull/10/commits/594350e94967e27aa3280a0871562e8ebc3e12d4, can this be merged now?

alexstrel committed:
moved dpl version since it needs some additional tuning

alexstrel commented:
@kakwok thanks for the update, looks good for me. (I moved dpl version into a separate feature branch since it still needs some tuning/testing before merge)

cerati commented:
based on comments above it looks like all is good, so I am merging
